ykcompile/src/lib.rs

```diff
@@ -94,77 +97,133 @@ impl CompiledTrace {
-            {
-                // We currently assume that we only trace a single function which leaves its return
-                // value in RAX. Since we now inline a function's return variable this won't happen
-                // automatically anymore. To keep things working, we thus copy the return value of
-                // the most outer function into RAX at the end of the trace.
-                dynasm!(self.asm
-                    ; mov rax, Rq(reg)
-                );
+        let is_rtn_var = local
+            == &self
+                .rtn_var
+                .as_ref()
+                .ok_or_else(|| CompileError::NoReturnLocal)?
+                .local;
+        match self.variable_location_map.get(local) {
+            Some(Location::Register(reg)) => {
+                // If this local is currently stored in a register, free it.
+                self.register_content_map.insert(*reg, None);
+                if is_rtn_var {
+                    // We currently assume that we only trace a single function which leaves its return
+                    // value in RAX. Since we now inline a function's return variable this won't happen
+                    // automatically anymore. To keep things working, we thus copy the return value of
+                    // the most outer function into RAX at the end of the trace.
+                    dynasm!(self.asm
+                        ; mov rax, Rq(reg)
+                    );
+                }
+            }
+            Some(Location::Stack(offset)) => {
+                if is_rtn_var {
+                    dynasm!(self.asm
+                        ; mov rax, [rsp + *offset]
+                    );
+                }
             }
-            self.available_regs.push(reg);
+            Some(Location::NotLive) => {}
+            None => {}
         }
-
+        self.variable_location_map.insert(*local, Location::NotLive);
         Ok(())
     }
 
     /// Copy the contents of the local `l2` into  `l1`.
     fn mov_local_local(&mut self, l1: Local, l2: Local) -> Result<(), CompileError> {
-        let lreg = self.local_to_reg(l1)?;
-        let rreg = self.local_to_reg(l2)?;
-        dynasm!(self.asm
-            ; mov Rq(lreg), Rq(rreg)
-        );
+        let lloc = self.local_to_location(l1)?;
+        let rloc = self.local_to_location(l2)?;
+        match (lloc, rloc) {
+            (Location::Register(lreg), Location::Register(rreg)) => {
+                dynasm!(self.asm
+                    ; mov Rq(lreg), Rq(rreg)
+                );
+            }
+            (Location::Register(reg), Location::Stack(off)) => {
+                dynasm!(self.asm
+                    ; mov Rq(reg), [rsp + off]
+                );
+            }
+            (Location::Stack(off), Location::Register(reg)) => {
+                dynasm!(self.asm
+                    ; mov [rsp + off], Rq(reg)
+                );
+            }
+            (Location::Stack(off1), Location::Stack(off2)) => {
```

Comment:
So if we didn't have RAX reserved, we'd kick another variable to the stack, but just temporarily?

Reply:
Yes, that would be the plan.